src/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs

```diff
@@ -156,6 +156,8 @@ public override bool TryComplete(Exception error)
             {
                 BoundedChannel<T> parent = _parent;
                 bool completeTask;
+                ReaderInteractor<bool> waitingReaders = parent._waitingReaders;
```

Comment:
This needs to be done inside the lock.  It should be moved to just before you set it to null, and it should then only be set to null if it's not already.

Reply:
>  It should be moved to just before you set it to null, and it should then only be set to null if it's not already.

Why don't we set it to null anyway without checking if it is already null?

Anyway, I'll do what you have suggested and you can comment more on the changes :-)